sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -144,16 +144,10 @@ class CacheManager extends Logging {
       } else {
         _.sameResult(plan)
       }
-    val plansToUncache = mutable.Buffer[CachedData]()
-    readLock {
-      val it = cachedData.iterator()
-      while (it.hasNext) {
-        val cd = it.next()
-        if (shouldRemove(cd.plan)) {
-          plansToUncache += cd
-        }
-      }
+    val cachedDataCopy = readLock {
+      cachedData.asScala.clone()
```

Comment:
.clone() is only making a shallow copy right? or it could be expensive.

Reply:
It is a shallow copy.  This definitely only makes a difference under very heavy load.  We haven't seen any impact in performance under smaller cases, and in our case it prevents the entire spark application from getting in a lock state where only one job gets to run at a time.